lib/std/http/Client.zig

```diff
@@ -1,62 +1,417 @@
+            .tls => return conn.tls_client.writeAll(conn.stream, buffer),
+        }
+    }
+
+    pub fn write(conn: *Connection, buffer: []const u8) !usize {
+        switch (conn.protocol) {
+            .plain => return conn.stream.write(buffer),
+            .tls => return conn.tls_client.write(conn.stream, buffer),
+        }
+    }
+};
+
 /// TODO: emit error.UnexpectedEndOfStream or something like that when the read
 /// data does not match the content length. This is necessary since HTTPS disables
 /// close_notify protection on underlying TLS streams.
 pub const Request = struct {
     client: *Client,
-    stream: net.Stream,
-    headers: std.ArrayListUnmanaged(u8) = .{},
-    tls_client: std.crypto.tls.Client,
-    protocol: Protocol,
-    response_headers: http.Headers = .{},
+    connection: Connection,
+    redirects_left: u32,
+    response: Response,
+    /// These are stored in Request so that they are available when following
+    /// redirects.
+    headers: Headers,
 
-    pub const Protocol = enum { http, https };
+    pub const Response = struct {
+        headers: Response.Headers,
+        state: State,
+        header_bytes_owned: bool,
+        /// This could either be a fixed buffer provided by the API user or it
+        /// could be our own array list.
+        header_bytes: std.ArrayListUnmanaged(u8),
+        max_header_bytes: usize,
 
-    pub const Options = struct {
+        pub const Headers = struct {
+            location: ?[]const u8 = null,
+            status: http.Status,
+            version: http.Version,
+            content_length: ?u64 = null,
+
+            pub fn parse(bytes: []const u8) !Response.Headers {
+                var it = mem.split(u8, bytes[0 .. bytes.len - 4], "\r\n");
+
+                const first_line = it.first();
+                if (first_line.len < 12)
+                    return error.ShortHttpStatusLine;
+
+                const version: http.Version = switch (int64(first_line[0..8])) {
+                    int64("HTTP/1.0") => .@"HTTP/1.0",
+                    int64("HTTP/1.1") => .@"HTTP/1.1",
+                    else => return error.BadHttpVersion,
+                };
+                if (first_line[8] != ' ') return error.InvalidHttpHeaders;
+                const status = @intToEnum(http.Status, parseInt3(first_line[9..12].*));
+
+                var headers: Response.Headers = .{
+                    .version = version,
+                    .status = status,
+                };
+
+                while (it.next()) |line| {
+                    var line_it = mem.split(u8, line, ": ");
+                    const header_name = line_it.first();
+                    const header_value = line_it.rest();
+                    if (std.ascii.eqlIgnoreCase(header_name, "location")) {
+                        headers.location = header_value;
+                    } else if (std.ascii.eqlIgnoreCase(header_name, "content-length")) {
```

Comment:
IIRC, if content-length is sent more than once you should use the first occurrence

Reply:
I do not see confirmation of this in https://www.rfc-editor.org/rfc/rfc2616

Reply:
Looks like my recollection was wrong looking at the standards

https://www.rfc-editor.org/rfc/rfc7230#section-3.3.2
> If a message is received that has multiple Content-Length header
>    fields with field-values consisting of the same decimal value, or a
>    single Content-Length header field with a field value containing a
>    list of identical decimal values (e.g., "Content-Length: 42, 42"),
>    indicating that duplicate Content-Length header fields have been
>    generated or combined by an upstream message processor, then the
>    recipient MUST either reject the message as invalid or replace the
>    duplicated field-values with a single valid Content-Length field
>    containing that decimal value prior to determining the message body
>    length or forwarding the message.
> 
>       Note: HTTP's use of Content-Length for message framing differs
>       significantly from the same field's use in MIME, where it is an
>       optional field used only within the "message/external-body"
>       media-type.

https://www.rfc-editor.org/rfc/rfc7230#section-3.3.3
> If a message is received without Transfer-Encoding and with
>        either multiple Content-Length header fields having differing
>        field-values or a single Content-Length header field having an
>        invalid value, then the message framing is invalid and the
>        recipient MUST treat it as an unrecoverable error